File: app/src/main/java/mono/hg/SettingsActivity.java
package mono.hg;

import android.content.Intent;
import android.os.Bundle;
import android.view.MenuItem;

import androidx.annotation.NonNull;
import androidx.appcompat.app.AppCompatActivity;
import androidx.appcompat.app.AppCompatDelegate;
import androidx.core.app.ActivityCompat;
import androidx.fragment.app.Fragment;
import androidx.fragment.app.FragmentTransaction;
import androidx.preference.Preference;
import androidx.preference.PreferenceFragmentCompat;

import mono.hg.helpers.PreferenceHelper;
import mono.hg.preferences.BasePreference;
import mono.hg.utils.ActivityServiceUtils;
import mono.hg.utils.Utils;
import mono.hg.utils.ViewUtils;
import mono.hg.wrappers.BackHandledFragment;

public class SettingsActivity extends AppCompatActivity
        implements BackHandledFragment.BackHandlerInterface, PreferenceFragmentCompat.OnPreferenceStartFragmentCallback {

    private BackHandledFragment selectedFragment;
    private CharSequence fragmentTitle;

    @Override public void onCreate(Bundle savedInstanceState) {
        if (!PreferenceHelper.hasEditor()) {
            PreferenceHelper.initPreference(this);
        }
        PreferenceHelper.fetchPreference();

        if (PreferenceHelper.getProviderList().isEmpty()) {
            Utils.setDefaultProviders(getResources());
        }

        // Check the caller of this activity.
        // If it's coming from the launcher itself, it will always have a calling activity.
        checkCaller();

        // Load the appropriate theme.
        switch (PreferenceHelper.appTheme()) {
            default:
            case "auto":
                AppCompatDelegate.setDefaultNightMode(AppCompatDelegate.MODE_NIGHT_AUTO_BATTERY);
                break;
            case "light":
                AppCompatDelegate.setDefaultNightMode(AppCompatDelegate.MODE_NIGHT_NO);
                break;
            case "dark":
                AppCompatDelegate.setDefaultNightMode(AppCompatDelegate.MODE_NIGHT_YES);
                setTheme(R.style.AppTheme_Dark);
                break;
            case "black":
                AppCompatDelegate.setDefaultNightMode(AppCompatDelegate.MODE_NIGHT_YES);
                break;
        }

        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_settings);

        if (getRequestedOrientation() != PreferenceHelper.getOrientation()) {
            setRequestedOrientation(PreferenceHelper.getOrientation());
        }

        if (getSupportActionBar() != null) {
            getSupportActionBar().setDisplayHomeAsUpEnabled(true);
        }

        if (savedInstanceState == null) {
            ViewUtils.setFragment(getSupportFragmentManager(), new BasePreference(), "settings");
        } else {
            fragmentTitle = savedInstanceState.getCharSequence("title");
            if (getSupportActionBar() != null) {
                getSupportActionBar().setTitle(fragmentTitle);
            }
        }
    }

    @Override protected void onSaveInstanceState(@NonNull Bundle outState) {
        super.onSaveInstanceState(outState);
        outState.putCharSequence("title", fragmentTitle);
    }

    @Override public void onBackPressed() {
        if (selectedFragment == null || !selectedFragment.onBackPressed()) {
            // Selected fragment did not consume the back press event.
            if (getSupportActionBar() != null) {
                if (getSupportFragmentManager().findFragmentByTag(
                        "settings") instanceof BasePreference) {
                    getSupportActionBar().setTitle(getString(R.string.title_activity_settings));
                } else {
                    getSupportActionBar().setTitle(fragmentTitle);
                }
            }
            super.onBackPressed();
        }
    }

    @Override public void setSelectedFragment(BackHandledFragment selectedFragment) {
        this.selectedFragment = selectedFragment;
    }

    @Override public boolean onOptionsItemSelected(MenuItem item) {
        if (item.getItemId() == android.R.id.home) {
            if (getSupportActionBar() != null) {
                if (getSupportFragmentManager().findFragmentByTag(
                        "settings") instanceof BasePreference) {
                    getSupportActionBar().setTitle(getString(R.string.title_activity_settings));
                } else {
                    getSupportActionBar().setTitle(fragmentTitle);
                }
            }
            super.onBackPressed();
            ActivityServiceUtils.hideSoftKeyboard(this);
            return true;
        }
        return super.onOptionsItemSelected(item);
    }

    public void checkCaller() {
        // If this activity is called from anywhere else but the launcher,
        // then the launcher needs to be informed of the changes made that it may not be aware of.
        if (getCallingActivity() == null && !PreferenceHelper.wasAlien()) {
            PreferenceHelper.isAlien(true);
        }
    }

    // Called when the activity needs to be restarted (i.e when a theme change occurs).
    // Allows for smooth transition between recreation.
    public void restartActivity() {
        Intent intent = getIntent();
        intent.addFlags(Intent.FLAG_ACTIVITY_NO_ANIMATION);
        intent.addFlags(Intent.FLAG_ACTIVITY_NO_HISTORY);
        intent.addFlags(Intent.FLAG_ACTIVITY_CLEAR_TOP | Intent.FLAG_ACTIVITY_NEW_TASK);
        ActivityCompat.finishAfterTransition(this);
        overridePendingTransition(android.R.anim.fade_in, android.R.anim.fade_out);
        startActivity(intent);
    }

    @Override public boolean onPreferenceStartFragment(PreferenceFragmentCompat caller, Preference pref) {
        // Instantiate the new Fragment
        final Fragment fragment = getSupportFragmentManager()
                .getFragmentFactory().instantiate(getClassLoader(), pref.getFragment());
        fragment.setTargetFragment(caller, 0);
        fragmentTitle = pref.getTitle();

        // Replace the existing Fragment with the new Fragment
        getSupportFragmentManager().beginTransaction()
                                   .setTransition(FragmentTransaction.TRANSIT_FRAGMENT_OPEN)
                                   .replace(R.id.fragment_container, fragment, pref.getKey())
                                   .addToBackStack(pref.getKey())
                                   .commit();

        // Update the Activity's action bar
        if (getSupportActionBar() != null) {
            getSupportActionBar().setTitle(fragmentTitle);
        }
        return true;
    }
}

File: app/src/main/java/mono/hg/SettingsActivity.kt
package mono.hg

import android.content.SharedPreferences
import android.content.res.Configuration
import android.content.res.Configuration.UI_MODE_NIGHT_NO
import android.os.Bundle
import android.view.MenuItem
import androidx.appcompat.app.AppCompatActivity
import androidx.appcompat.widget.Toolbar
import androidx.fragment.app.FragmentTransaction
import androidx.preference.Preference
import androidx.preference.PreferenceFragmentCompat
import com.google.android.material.progressindicator.LinearProgressIndicator
import mono.hg.databinding.ActivitySettingsBinding
import mono.hg.helpers.PreferenceHelper
import mono.hg.preferences.BasePreference
import mono.hg.utils.ActivityServiceUtils
import mono.hg.utils.Utils
import mono.hg.utils.ViewUtils
import mono.hg.utils.compatHide

/**
 * Activity hosting all of preference fragments.
 *
 * This activity can be called through 'Additional setting' in the System settings as well.
 */
class SettingsActivity : AppCompatActivity(),
    PreferenceFragmentCompat.OnPreferenceStartFragmentCallback,
    SharedPreferences.OnSharedPreferenceChangeListener {
    private var fragmentTitle: CharSequence? = null
    private lateinit var binding: ActivitySettingsBinding
    private lateinit var toolbar: Toolbar
    lateinit var progressBar: LinearProgressIndicator

    // The list of preferences that will trigger a launcher restart.
    // This restart will be done in LauncherActivity itself.
    private val preferenceRestartTrigger =
        "app_theme app_accent widget_space_visible app_list_mode list_bg icon_hide_switch adaptive_shade_switch icon_pack list_order shade_view_switch"

    public override fun onCreate(savedInstanceState: Bundle?) {
        if (! PreferenceHelper.hasEditor()) {
            PreferenceHelper.initPreference(this)
        }

        PreferenceHelper.fetchPreference()
        if (PreferenceHelper.providerList.isEmpty()) {
            PreferenceHelper.updateProvider(Utils.setDefaultProviders(resources, ArrayList()))
        }

        // Check the caller of this activity.
        // If it's coming from the launcher itself, it will always have a calling activity.
        checkCaller()
        setActivityTheme()

        if (requestedOrientation != PreferenceHelper.orientation) {
            requestedOrientation = PreferenceHelper.orientation
        }

        binding = ActivitySettingsBinding.inflate(layoutInflater)
        setContentView(binding.root)

        super.onCreate(savedInstanceState)

        toolbar = binding.toolbar
        progressBar = binding.progressBar.apply {
            setIndicatorColor(PreferenceHelper.darkerAccent)
            trackColor = PreferenceHelper.accent
            compatHide()
        }

        setSupportActionBar(toolbar)
        supportActionBar?.setDisplayHomeAsUpEnabled(true)

        if (savedInstanceState == null) {
            supportActionBar?.title = getString(R.string.title_activity_settings)
            ViewUtils.setFragment(supportFragmentManager, BasePreference(), "settings")
        } else {
            fragmentTitle = savedInstanceState.getCharSequence("title")
                ?: getString(R.string.title_activity_settings)
            supportActionBar?.title = fragmentTitle
        }
    }

    override fun onResume() {
        super.onResume()
        PreferenceHelper.preference.registerOnSharedPreferenceChangeListener(this)
    }

    override fun onPause() {
        super.onPause()
        PreferenceHelper.preference.unregisterOnSharedPreferenceChangeListener(this)
    }

    override fun onSharedPreferenceChanged(pref: SharedPreferences?, key: String?) {
        if (preferenceRestartTrigger.contains(key.toString())) {
            PreferenceHelper.update("require_refresh", true)

            // A reinitialisation isn't necessary since we're doing a full refresh.
            PreferenceHelper.update("require_reinit", false)
        } else {
            // These preferences will only require a refresh of PreferenceHelper.
            PreferenceHelper.update("require_reinit", true)
        }
    }

    override fun onSaveInstanceState(outState: Bundle) {
        super.onSaveInstanceState(outState)
        outState.putCharSequence("title", fragmentTitle)
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        if (item.itemId == android.R.id.home) {
            super.onBackPressed()
            ActivityServiceUtils.hideSoftKeyboard(this, currentFocus)
            return true
        }
        return super.onOptionsItemSelected(item)
    }

    private fun checkCaller() {
        // If this activity is called from anywhere else but the launcher,
        // then the launcher needs to be informed of the changes made that it may not be aware of.
        if (callingActivity == null && ! PreferenceHelper.wasAlien()) {
            PreferenceHelper.isAlien(true)
        }
    }

    private fun setActivityTheme() {
        ViewUtils.switchTheme(this, false)
        if (Utils.atLeastLollipop() && resources.configuration.uiMode and
            Configuration.UI_MODE_NIGHT_MASK == UI_MODE_NIGHT_NO
        ) {
            window.statusBarColor = PreferenceHelper.darkerAccent
        }
    }

    override fun onPreferenceStartFragment(
        caller: PreferenceFragmentCompat,
        pref: Preference
    ): Boolean {
        // Instantiate the new Fragment
        val fragment =
            supportFragmentManager.fragmentFactory.instantiate(classLoader, pref.fragment)
        fragment.setTargetFragment(caller, 0)
        fragmentTitle = pref.title

        // Replace the existing Fragment with the new Fragment
        supportFragmentManager.beginTransaction()
            .setTransition(FragmentTransaction.TRANSIT_FRAGMENT_OPEN)
            .replace(R.id.fragment_container, fragment, pref.key)
            .addToBackStack(pref.key)
            .commit()

        // Update the Activity's action bar
        supportActionBar?.title = fragmentTitle
        return true
    }
}